### PR TITLE
Port experimental base_turf system from Baystation again

### PR DESCRIPTION
**Pull request by ghost**
Fuck you too for not even trying when you were working on Endgame @N3X15

Also thanks for fucking up my PR by stealth-merging a fix @Comic

Because Baystation has finally gotten around to it, and because the code looked simple, I decided to get my hands on porting a system that allows up to define base_turfs. As a bonus, it also defines them per Z-level. So if you wanted, you could have a space station, and a Z-level on a map-spanning asteroid, and then an away mission on land, all in the same file

Note that you can modify it mid-round via a DEBUG flag verb, it will however only affect destroyed turfs (it won't change space, which I suppose is intentional)

Tested were destroying walls and floors and even dropping a bomb in the middle of Medbay

Not tested were more finicky conditions like the Singularity or buildmode. I simply replaced all ChangeTurf() instructions that pointed to space, because that is how Baystation implemented it

Only problem with this is that it doesn't work on a per-map basis, so unless some wizardry is done on that code, it'll be hard to get anything interesting going with it

Discussions go below. Since this is pure code refactor, I doubt there'll be much up to debate

Old PR over at #5167

Example of the system working live :

dune

### CODE REVIEW

**Comment by PJB3005** [comment]
Fucking learn how to rebase already gee

File: code/game/turfs/turf.dm [hunk 1]
```diff
@@ -544,7 +547,7 @@
 
 
 /turf/proc/cultify()
-	ChangeTurf(/turf/space)
+	ChangeTurf(get_base_turf(src.z))
```
Comment (clusterfack) [inline code/game/turfs/turf.dm]: What if the base turf is the turf being cultified?

File: code/game/turfs/turf.dm [hunk 2]
```diff
@@ -426,8 +429,8 @@
 */
 
 /turf/proc/ReplaceWithLattice()
-	src.ChangeTurf(/turf/space)
-	new /obj/structure/lattice( locate(src.x, src.y, src.z) )
+	src.ChangeTurf(get_base_turf(src.z))
+	new /obj/structure/lattice(locate(src.x, src.y, src.z))
```
Comment (clusterfack) [inline code/game/turfs/turf.dm]: Locate unnecessary, use src

File: code/game/turfs/turf.dm [hunk 3]
```diff
@@ -554,7 +557,7 @@
 				continue
 			if(O.invisibility == 101)
 				O.singularity_act()
-	ChangeTurf(/turf/space)
+	ChangeTurf(get_base_turf(src.z))
```
Comment (clusterfack) [inline code/game/turfs/turf.dm]: Same problem as cultify

File: code/game/turfs/turf.dm (hunk 1 above)
Comment (dylanstrategie) [inline code/game/turfs/turf.dm]: Good question

File: code/modules/bomberman/bomberman.dm [hunk 4]
```diff
@@ -1165,7 +1165,7 @@ var/global/list/arenas = list()
 		for(var/atom/movable/AM in T)
 			AM.areaMaster = get_area_master(T)
 		if(open_space && (under.name == "Space"))
```
Comment (clusterfack) [inline code/modules/bomberman/bomberman.dm]: This one really does seem to explicitly want space, doesn't it?

File: code/game/turfs/turf.dm (hunk 1 above)
Comment (PJB3005) [inline code/game/turfs/turf.dm]: er, it'll cultify to that turf, which'll cultify to that turf...

it won't be an infinite loop but it'll go on for infinity

File: code/game/turfs/turf.dm (hunk 3 above)
Comment (PJB3005) [inline code/game/turfs/turf.dm]: same, it'll keep eating it for infinity

@dylanstrategie, you could make the singuloth not eat the base turf

File: code/modules/bomberman/bomberman.dm (hunk 4 above)
Comment (dylanstrategie) [inline code/modules/bomberman/bomberman.dm]: Blame Deity, I'll give it a look

File: code/game/objects/items/weapons/RCD.dm [hunk 5]
```diff
@@ -221,7 +221,7 @@ RCD
 					if(do_after(user,A, 50))
 						if(!useResource(decon_cost, user)) return 0
 						activate()
-						A:ChangeTurf(A:under_turf)
+						A:ChangeTurf(get_base_turf(A.z))
```
Comment (clusterfack) [inline code/game/objects/items/weapons/RCD.dm]: That feel when I deconstruct a base turf and it gives me back the same base turf.

Comment (dylanstrategie) [inline code/game/objects/items/weapons/RCD.dm]: Can't fix that, that would be adminbuse (if the base_turf was made to be floor or wall, respectively)

File: code/modules/bomberman/bomberman.dm (hunk 4 above)
Comment (dylanstrategie) [inline code/modules/bomberman/bomberman.dm]: After checking, under.name seems to be kind of a blueprint for building the arena, I'll just look to see if I should change it to say Base Turf

Comment (dylanstrategie) [inline code/modules/bomberman/bomberman.dm]: Meh, not an issue, it's just used to plan and draw the arena

File: code/game/turfs/turf.dm (hunk 3 above)
Comment (dylanstrategie) [inline code/game/turfs/turf.dm]: This was fixed

File: code/__HELPERS/unsorted.dm [hunk 6]
```diff
@@ -990,7 +990,7 @@ var/list/ignored_keys = list("loc", "locs", "parent_type", "vars", "verbs", "typ
 							T.icon = 'icons/turf/snow.dmi'
 							T.icon_state = "snow"
 						else
-							T.ChangeTurf(/turf/space)
+							T.ChangeTurf(get_base_turf(T.z))
 							switch(universe.name)	//for some reason using OnTurfChange doesn't actually do anything in this case.
```
Comment (clusterfack) [inline code/__HELPERS/unsorted.dm]: This applies space overlays

File: code/game/turfs/turf.dm [hunk 7]
```diff
@@ -426,8 +429,8 @@
 */
 
 /turf/proc/ReplaceWithLattice()
-	src.ChangeTurf(/turf/space)
-	new /obj/structure/lattice( locate(src.x, src.y, src.z) )
+	src.ChangeTurf(get_base_turf(src.z))
+	new /obj/structure/lattice(src)
```
Comment (clusterfack) [inline code/game/turfs/turf.dm]: if(istype(space,src))
new lattice

Comment (dylanstrategie) [inline code/game/turfs/turf.dm]: Will do for now I suppose

File: code/__HELPERS/unsorted.dm (hunk 6 above)
Comment (dylanstrategie) [inline code/__HELPERS/unsorted.dm]: I'll limit it to spess until someone unsnowflakes them, then

**Comment by dylanstrategie** [comment]
Tested and ready to rock

**Pull request closed without merging**